resolver/resolver.cc

```diff
@@ -1447,28 +1447,26 @@ class ResolveSignaturesWalk {
     void injectOptionalArgs(core::MutableContext ctx, ast::MethodDef *mdef) {
         ast::InsSeq::STATS_store lets;
 
-        if (mdef->symbol.data(ctx)->isAbstract()) {
-            // If we checked that abstract methods don't have defined bodies earlier (currently done in infer)
-            // we could also use this technique to check default arguments of abstract methods.
-            return;
-        }
-
         int i = -1;
         for (auto &argSym : mdef->symbol.data(ctx)->arguments()) {
             i++;
             auto &argExp = mdef->args[i];
             auto argType = argSym.type;
 
             if (auto *optArgExp = ast::cast_tree<ast::OptionalArg>(argExp.get())) {
-                // Using optArgExp's loc will make errors point to the arg list, even though the T.let is in the
-                // body.
-                auto let = make_unique<ast::Cast>(optArgExp->loc, argType, optArgExp->default_->deepCopy(),
-                                                  core::Names::let());
-                lets.emplace_back(std::move(let));
+                if (!argType) {
+                    optArgExp->default_ = nullptr;
+                } else {
+                    // Using optArgExp's loc will make errors point to the arg list, even though the T.let is in the
+                    // body.
+                    auto let =
+                        make_unique<ast::Cast>(optArgExp->loc, argType, move(optArgExp->default_), core::Names::let());
+                    lets.emplace_back(std::move(let));
+                }
             }
         }
 
-        if (!lets.empty()) {
+        if (!lets.empty() && !mdef->symbol.data(ctx)->isAbstract()) {
```

Comment:
Could you leave a comment here about why this function removes defaults from abstract methods without adding them to the method body?

Reply:
Sure! It is because we aren't going to do anything with them anyways, so leaving them in the tree is just confusing. They are 'weirdly' in the tree in that we don't typecheck them but they still sit there so I thought the invariant of never having anything in the `default_` after this phase was easier to reason about.